src/fns.ts

```diff
@@ -138,32 +171,67 @@ export const throughputLimiter = (
    * Call before processing a batch of units. After the first call, a subsequent
    * call assumes that the `numUnits` from the previous call were processed. A
    * call to `throttle` may sleep for a given period of time depending on
-   * `maxUnitsPerSec` and the total number of units over the current window.
+   * `maxUnitsPerPeriod` and the total number of units over the current window.
    */
-  const throttle = async (numUnits: number) => {
-    debugTL('throttle called - %d', numUnits)
+  const throttle = async () => {
     // Skip check if maxUnitsPerSec is Infinity
-    if (maxUnitsPerSec === Infinity) {
+    if (maxUnitsPerPeriod === Infinity) {
       debugTL('exiting throttle - maxUnitsPerSec is Infinity')
       return
     }
     // Sleep if the current rate is above the max allowed. Repeat
     // until the rate has dropped sufficiently.
-    while (getCurrentRate() > maxUnitsPerSec) {
+    while (getCurrentRate() > maxUnitsPerPeriod) {
       debugTL('sleeping for %d', sleepTime)
       await sleep(sleepTime)
     }
-    slidingWindow.push({ timestamp: new Date().getTime(), numUnits })
-    if (slidingWindow.length > windowLength) {
-      debugTL('truncating slidingWindow')
-      slidingWindow.shift()
+  }
+
+  const append = (numUnits: number) => {
+    debugTL('append called - %d', numUnits)
+    const now = new Date().getTime()
+    // Add the current invocation to the sliding window
+    slidingWindow.push({ timestamp: now, numUnits })
+    // Truncate the sliding window according to the window length
+    if (slidingWindow.length > maxWindowLength) {
+      const shifted = slidingWindow.shift()
+      debugTL('removed due to length: %o', shifted)
+    }
+    // Remove expired invocations
+    if (expireAfter !== Infinity) {
+      // Remove invocations that are older than expireAfter
+      while (now - slidingWindow[0]?.timestamp > expireAfter) {
+        const shifted = slidingWindow.shift()
+        debugTL('removed expired: %o', shifted)
+      }
     }
     debugTL('slidingWindow %o', slidingWindow)
   }
 
+  /**
+   * This method is a combination of `throttle` and `append`. It will throttle
+   * first and then append the number of units to the sliding window.
+   */
+  const throttleAndAppend = async (numUnits: number) => {
+    await throttle()
+    append(numUnits)
+  }
+
+  /**
+   * This method is a combination of `append` and `throttle`. It will append
+   * the number of units to the sliding window and then throttle.
+   */
+  const appendAndThrottle = async (numUnits: number) => {
+    append(numUnits)
+    await throttle()
+  }
```

Comment:
I like how you've separated this out -- it's more flexible and easier to understand.